### PR TITLE
[lte][agw] Adding corruption of state AGW S1AP test

### LINKED ISSUE #3933: [lte][agw] State recovery service: Corrupt state for each service
Add S1AP tests that involve corrupting state on redis to validate health service for stateless AGW

### PR DESCRIPTION
**Pull request by ardzoht**
Signed-off-by: Alejandro Rodriguez <alexrod@fb.com>

## Summary

- Adding test_attach_detach_with_corrupt_agw_state.py that validates state corruption health recovery process
- Minor change for state_cli.py with python3 shebang
- Updating health.yml to threshold 2 for service restarts

## Test Plan

- make integ_test
- Running added s1ap test

```
**********************TPT open server******** 0

 Server Socket FD =[13]
************* Testing *Time elapsed(in usec): for timer expiry               :1288 usec
AGW is stateless
APN Correction configured
Using subscriber IMSI IMSI001010000000001
Using subscriber IMSI IMSI001010000000002
************************* UE device config for ue_id  1
************************* UE device config for ue_id  2
************************* Configuring IP block
************************* Waiting for IP changes to propagate
************************* S1 setup
************************* UE App config
************************* Running End to End attach for  UE id  1
ue id 1 found

*************NbUiNbuHdlUeRadioCapMsg: Send UE cap msg SUCCESS*********
PDN_TYPE: 1
IPv4 PDN_ADDRESS:192 168 128 41
************************* Corrupting MME state
Corrupted mme_nas_state on redis
************************* Restarting MME service on gateway
Waiting for 0 seconds
Waiting for 1 seconds
Waiting for 2 seconds
Waiting for 3 seconds
Waiting for 4 seconds
Waiting for 5 seconds
Waiting for 6 seconds
Waiting for 7 seconds
Waiting for 8 seconds
Waiting for 9 seconds
Waiting for 10 seconds
Waiting for 11 seconds
Waiting for 12 seconds
Waiting for 13 seconds
Waiting for 14 seconds
Waiting for 15 seconds
Waiting for 16 seconds
Waiting for 17 seconds
Waiting for 18 seconds
Waiting for 19 seconds
Waiting for 20 seconds
Waiting for 21 seconds
Waiting for 22 seconds
Waiting for 23 seconds
Waiting for 24 seconds
Waiting for 25 seconds
Waiting for 26 seconds
Waiting for 27 seconds
Waiting for 28 seconds
Waiting for 29 seconds
Waiting for 30 seconds
Waiting for 31 seconds
Waiting for 32 seconds
Waiting for 33 seconds
Waiting for 34 seconds
Waiting for 35 seconds
Waiting for 36 seconds
Waiting for 37 seconds
Waiting for 38 seconds
Waiting for 39 seconds
Waiting for 40 seconds
Waiting for 41 seconds
Waiting for 42 seconds
Waiting for 43 seconds
Waiting for 44 seconds
Waiting for 45 seconds
Waiting for 46 seconds
Waiting for 47 seconds
Waiting for 48 seconds
Waiting for 49 seconds
Waiting for 50 seconds
Waiting for 51 seconds
Waiting for 52 seconds
Waiting for 53 seconds
Waiting for 54 seconds
Waiting for 55 seconds
Waiting for 56 seconds
Waiting for 57 seconds
Waiting for 58 seconds
Waiting for 59 seconds
Waiting for 60 seconds
Waiting for 61 seconds
Waiting for 62 seconds
Waiting for 63 seconds
Waiting for 64 seconds
Waiting for 65 seconds
Waiting for 66 seconds
Waiting for 67 seconds
Waiting for 68 seconds
Waiting for 69 seconds
Waiting for 70 seconds
Waiting for 71 seconds
Waiting for 72 seconds
Waiting for 73 seconds
Waiting for 74 seconds
Waiting for 75 seconds
Deleting Ue Context from Traffic Handler
Waiting for 76 seconds
Waiting for 77 seconds
Waiting for 78 seconds
Waiting for 79 seconds
Waiting for 80 seconds
Waiting for 81 seconds
Waiting for 82 seconds
Waiting for 83 seconds
Waiting for 84 seconds
Waiting for 85 seconds
Waiting for 86 seconds
Waiting for 87 seconds
Waiting for 88 seconds
Waiting for 89 seconds
Waiting for 90 seconds
Waiting for 91 seconds
Waiting for 92 seconds
Waiting for 93 seconds
Waiting for 94 seconds
Waiting for 95 seconds
Waiting for 96 seconds
Waiting for 97 seconds
Waiting for 98 seconds
Waiting for 99 seconds
************************* S1 setup
************************* Re-running End to End attach for  UE id  1
ue id 1 found
PDN_TYPE: 1
IPv4 PDN_ADDRESS:192 168 128 41

*************NbUiNbuHdlUeRadioCapMsg: Send UE cap msg SUCCESS*********
************************* Running UE detach for UE id  1
Deleting Ue Context from Traffic Handler
************************* send SCTP SHUTDOWN
.
----------------------------------------------------------------------
Ran 1 test in 105.144s

OK
sleep 1
```

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/configs/health.yml [hunk 1]
```diff
@@ -21,7 +21,7 @@ state_recovery:
     - pipelined
 
   # Number of restarts of services_check that triggers recovery process
-  restart_threshold: 3
+  restart_threshold: 2
```
Comment (ulaskozat) [inline lte/gateway/configs/health.yml:24]: what is the reason for this change?

File: lte/gateway/python/integ_tests/s1aptests/test_attach_detach_with_corrupt_agw_state.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+            self._s1ap_wrapper.magmad_util.exec_command(
+                "sudo service sctpd restart"
+            )
+
+            for j in range(100):
+                print("Waiting for", j, "seconds")
+                time.sleep(1)
+
+            # Re-establish S1 connection between eNB and MME
+            self._s1ap_wrapper._s1setup()
```
Comment (ulaskozat) [inline lte/gateway/python/integ_tests/s1aptests/test_attach_detach_with_corrupt_agw_state.py]: why are we restarting sctpd? I thought the idea was to restart MME.

File: lte/gateway/configs/health.yml (hunk 1 above)
Comment (ardzoht) [reply lte/gateway/configs/health.yml:24]: From the testing we've done we are tuning this to trigger clearing of corrupt state if there are more than 2 restarts over a 60sec period.

**Review by ssanadhya — APPROVED** [review]

File: lte/gateway/python/integ_tests/s1aptests/test_attach_detach_with_corrupt_agw_state.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+import time
+import unittest
+
+import s1ap_types
+import s1ap_wrapper
+from s1ap_utils import MagmadUtil
+
+
+class TestAttachDetachWithCorruptMMEState(unittest.TestCase):
```
Comment (ulaskozat) [inline lte/gateway/python/integ_tests/s1aptests/test_attach_detach_with_corrupt_agw_state.py]: File name and test name are not matching. Also it would be good to include stateless in the name as in other stateless tests.

**Review by ulaskozat — APPROVED** [review]
LGTM, but change the file name in line with the test name and also modify the test name to reflect that this is a stateless test.

File: lte/gateway/python/integ_tests/s1aptests/test_attach_detach_with_corrupt_stateless_mme.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+            # Wait on EMM Information from MME
+            self._s1ap_wrapper._s1_util.receive_emm_info()
+
+            print("************************* Corrupting %s state" % s)
+            self._s1ap_wrapper.magmad_util.corrupt_agw_state(
+                services_state_dict[s])
+
+            print("************************* Restarting MME service on",
+                  "gateway")
+            self._s1ap_wrapper.magmad_util.restart_services(["mme"])
```
Comment (pshelar) [inline lte/gateway/python/integ_tests/s1aptests/test_attach_detach_with_corrupt_stateless_mme.py]: can you use the iterator variable s.key

File: lte/gateway/python/integ_tests/s1aptests/test_attach_detach_with_corrupt_stateless_mme.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+            for j in range(100):
+                print("Waiting for", j, "seconds")
```
Comment (pshelar) [inline lte/gateway/python/integ_tests/s1aptests/test_attach_detach_with_corrupt_stateless_mme.py:66]: can we check if MME is in correct state in the loop rather than waiting for 100 sec?

Comment (ardzoht) [reply lte/gateway/python/integ_tests/s1aptests/test_attach_detach_with_corrupt_stateless_mme.py:66]: From S1AP tester it becomes difficult to check for health / status of services, this is possible but would require refactoring and sending remote commands to the gateway VM to get service's state, I'd want to keep it as that since it's a defined time to wait for the service to become healthy.

**Review by pshelar — APPROVED** [review]
